dispatch/dispatch.go

```diff
@@ -0,0 +1,351 @@
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sqs"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/go-kit/kit/metrics"
+	"github.com/xmidt-org/mimisbrunnr/model"
+	"github.com/xmidt-org/webpa-common/logging"
+	"github.com/xmidt-org/webpa-common/xhttp"
+	"github.com/xmidt-org/wrp-go/v2"
+	// "github.com/xmidt-org/wrp-go/wrp"
+)
+
+const (
+	defaultMinQueueSize = 5
+	minMaxWorkers       = 5
+)
+
+type D interface {
+	Start(context.Context) error
+	Dispatch(deviceID string, msg *wrp.Message) error
+	Update(norn model.Norn)
+	Stop(context.Context) error
+}
+
+// failureText is human readable text for the failure message
+const FailureText = `Unfortunately, your endpoint is not able to keep up with the ` +
+	`traffic being sent to it.  Due to this circumstance, all notification traffic ` +
+	`is being cut off and dropped for a period of time.  Please increase your ` +
+	`capacity to handle notifications, or reduce the number of notifications ` +
+	`you have requested.`
+
+func (d Dispatcher) Start(_ context.Context) error {
+	sess, err := session.NewSession(&aws.Config{
+		Region:      aws.String(d.Norn.Destination.AWSConfig.Sqs.Region),
+		Credentials: credentials.NewStaticCredentials(d.Norn.Destination.AWSConfig.ID, d.Norn.Destination.AWSConfig.AccessKey, d.Norn.Destination.AWSConfig.SecretKey),
+	})
+	if err != nil {
+		return err
+	}
+	d.SqsClient = sqs.New(sess)
+
+	d.Wg.Add(1)
+	go d.sendEvents()
+	return nil
+
+}
+
+func (d Dispatcher) Dispatch(deviceID string, msg *wrp.Message) error {
+	if deviceID == d.Norn.DeviceID {
+		en := NewEventWithID(msg, deviceID)
+		select {
+		case d.DispatchQueue.Load().(chan *eventWithID) <- en:
+			d.Measures.EventQueueDepthGauge.Add(1.0)
+		default:
+			d.queueOverflow()
+			d.Measures.DroppedQueueCount.Add(1.0)
+		}
+	}
+	return nil
+}
+
+func (d Dispatcher) sendEvents() {
+	defer d.Wg.Done()
+	queue := d.DispatchQueue.Load().(chan *eventWithID)
+	select {
+
+	case en, ok := <-queue:
+		if !ok {
+			break
+		}
+		d.Measures.EventQueueDepthGauge.Add(-1.0)
+
+		d.Mutex.RLock()
+		deliverUntil := time.Unix(0, d.Norn.ExpiresAt)
+		dropUntil := d.DropUntil
+		d.Mutex.RUnlock()
+
+		now := time.Now()
+
+		if now.Before(dropUntil) {
+			d.Measures.DroppedCutoffCounter.Add(1.0)
+		}
+		if now.After(deliverUntil) {
+			d.Measures.DroppedExpiredCounter.Add(1.0)
+		}
+
+		d.Workers.Acquire()
+		d.Measures.WorkersCount.Add(1.0)
+		go d.send(en.Event)
+	}
+	for i := 0; i < d.MaxWorkers; i++ {
+		d.Workers.Acquire()
+	}
+
+	return
+
+}
+
+func (d *Dispatcher) empty(droppedCounter metrics.Counter) {
+	droppedMsgs := d.DispatchQueue.Load().(chan *eventWithID)
+	d.DispatchQueue.Store(make(chan *wrp.Message, d.QueueSize))
+	droppedCounter.Add(float64(len(droppedMsgs)))
+	d.Measures.EventQueueDepthGauge.Set(0.0)
+	return
+}
+
+// called to deliver event
+func (d Dispatcher) send(msg *wrp.Message) {
+	defer d.Wg.Done()
```

Comment:
defer is only one function/line:
https://blog.golang.org/defer-panic-and-recover

To defer multiple things in the same function, you have to create a function and put all the calls in it:
```
defer func(){
    firstThing()
    secondThing()
}